Prepare a workflow for publishing releases

nirinchev opened the pull request:
This implements a workflow that uploads binaries to S3 and optionally to pub.dev. It roughly does the following:

* Prepare packages:
  1. Downloads binaries from PR build
  1. Infers the version
    a. [Prod] version is equal to whatever is in pubspec.yaml
    a. [Staging] version is equal to whatever is in pubspec.yaml + commit sha
  1. Runs the `archive` command to produce the platform-specific archives
  1. Updates all references to `realm_common` and `realm_generator`
    a. [Prod] those point to the upcoming version
    a. [Staging] those point to `../realm_common` and `../realm_generator`
  1. Removes `publish_to: none ` from all packages
  1. Prepares the various packages in their respective folders and runs `dart pub publish --dry run` on those
  1. Uploads the release folder containing all the packages as well as all the binaries to S3 (`static.realm.io/downloads/dart/version`)
* Publish packages (only prod):
  1. Downloads the release folder
  1. Uploads all packages to pub.dev
  1. Merges the existing release PR
  1. Publishes a github release
  1. Posts to #realm-releases

This introduces some changes to `install_command.dart` and `init.dart`:
1. Reworked `install_command` to remove most of the code duplication
2. Removed some unused arguments from init.dart to make it more obvious where locations are
3. `install_command` now works with file references. It will still attempt to download binaries for those unless `publish_to: none` is specified in the reference pubspec.yaml
4. `install_command` now downloads packages from S3 rather than github releases.

Example S3 url for the packages: https://s3.amazonaws.com/static.realm.io/downloads/dart/0.3.0-37c7a6092191f97dcff467dca4bfe8be9405d77c/packages.tar.gz

Example S3 url for the binaries: https://s3.amazonaws.com/static.realm.io/downloads/dart/0.3.0-37c7a6092191f97dcff467dca4bfe8be9405d77c/macos.tar.gz

Example Github release: https://github.com/realm/realm-dart/releases/tag/untagged-a796952aac036395c582

Example slack post: https://mongodb.slack.com/archives/CMMLKL5SB/p1651685930973989

TODO:
- [x] In ci.yml: update `todo-update-me-to-release` to `release`
- [x] In prepare-release.yml: update `test-master` and `test-release` to `master` and `release`.
- [x] In publish-release.yml: Remove --dry-run from `Publish packages to pub.dev` step
- [x] In publish-release.yml: Update `Publish Github Release.commit` to `master`
- [x] In publish-release.yml: Update `Publish Github Release.draft` to `false`

Follow-up tasks:
- [ ] Update ci.yml to parallelize all architectures once https://github.com/realm/realm-dart/pull/514 is merged
- [ ] Prepare a workflow to update the standard text of changelog.md after merging (it's a bit of an annoyance so not worth delaying this PR).
- [ ] Update realm/realm_dart references in pubspec.ymls to also point to the released version (these are the two examples).

nielsenko reviewed (approved):
LGTM

blagoev commented on .github/workflows/publish-release.yml:
```diff
@@ -3,22 +3,227 @@ on:
+      with:
+        branch: ${{ github.ref }}
+
+    - name: Merge Pull Request
```
Should these steps be checking a success status of something before it. Unsure what will happen if we fail to upload binaries to S3 and we merge to master this the release PR

nirinchev replied:
The job fails it any of its steps fails, so we won't get this far if something earlier fails.

blagoev commented on .github/workflows/publish-release.yml:
```diff
@@ -3,22 +3,227 @@ on:
+        number: ${{ steps.find-pull-request.outputs.number }}
+        method: squash
+
+    - name: Publish Github Release
```
The same here. Should we be executing these steps only if upload binaries succeeded.

blagoev commented on lib/src/cli/install/install_command.dart:
```diff
@@ -237,195 +125,82 @@ class InstallCommand extends Command<void> {
   }
 
-  Future<String> getRealmPackagePath() async {
+  Future<String> _getRealmPackagePath() async {
```
Making all these methods and fields private only hurts readability imo. There no reason to hide these since we are not using this class anywhere else so we need to care of its public interface.

nirinchev replied:
I agree - I was wondering what the best practice was around this. The main reason why I did it was that there was some unused method that wasn't flagged by the compiler because it was public. But now that they're all used, I can make them public again.

blagoev commented on lib/src/cli/install/install_command.dart:
```diff
@@ -237,195 +125,82 @@ class InstallCommand extends Command<void> {
-    if (await skipInstall()) {
-      print("Realm install command started from within the realm-dart repo or one of the realm packages is a path dependency or has a path dependency override."
-          " Skipping install.");
+    if (realmPubspec.publishTo == 'none') {
```
The previous implementation checked the package names in skipInstall as well. This one only checks publish to for skiping install. This makes this susceptible in the future the check to fail if someone forgets and commits in any realm package the `pubspec.yaml` file without `publish_to: none`. 
I suggest we keep the `skipInstall` implementation which also checked the package name

blagoev replied:
Also I am not sure why are we allowing path dependencies for `realm` and `realm_dart` packages. What is the benefit here. Imo we should be checking for that as the previous version since this is most certainly a bug or we are the ones executing the install command. In such cases install should not happen

nirinchev replied:
The idea is to simplify using prerelease binaries here. Since we're not uploading those to pub.dev, the only other way for people to reference it is as a path dependency. I'd rather have us support that scenario, even if it's for internal use. It's a lot simpler to tell docs/product to download and unzip a package and then run install than to tell them where to unzip the binaries for each platform.

nirinchev replied:
I can bring back the package name check - not sure it brings a whole lot value, but it doesn't hurt either.

blagoev commented on lib/src/cli/archive/archive_command.dart:
```diff
@@ -47,12 +47,10 @@ class ArchiveCommand extends Command<void> {
-    if (options.outputFile == null) {
-      abort("output-file option not specified");
-    }
+    final outputFile = options.outputFile ?? '${options.sourceDir}.tar.gz';
```
Output file is marked as required. So we should be validating it here as before. Or if we really want to make it optional than we need to change `options.dart`. Imo being explicit here only helps, cause this command is an internal one and not something users will be using. So being explicit here only helps us - where goes what when reading the various cmake/gh actions scripts.

blagoev commented on lib/src/cli/install/install_command.dart:
```diff
@@ -237,195 +125,82 @@ class InstallCommand extends Command<void> {
-      print("Realm install command started from within the realm-dart repo or one of the realm packages is a path dependency or has a path dependency override."
-          " Skipping install.");
+    if (realmPubspec.publishTo == 'none') {
+      print("Referencing $_packageName@${realmPubspec.version} which hasn't been published. Skipping download.");
```
This error message is misleading. When we publish a version we are going to keep the `publish_to: none` in the repo.

nirinchev replied:
This is intentional - we don't want to download binaries if we're working with the dart repo. When we upload the packages to S3, we remove the `publish_to: none`, so if you download a package from S3, you can reference it by path and run the install command to download the binaries.

blagoev replied:
Not sure what do you mean. How are we uploading the packages to S3. `dart pub publish` does not do that.

blagoev replied:
I added a discussion point for this. I think we can handle this use case better

nirinchev commented on flutter/realm_flutter/linux/CMakeLists.txt:
```diff
@@ -37,7 +37,7 @@ include(${EPHEMERAL_DIR}/generated_config.cmake)
   OUTPUT_VARIABLE output
   RESULT_VARIABLE result
-  COMMAND_ERROR_IS_FATAL ANY
+  # COMMAND_ERROR_IS_FATAL ANY
```
@blagoev and @fealebenpae - I commented this out to allow me to reduce the minimum required cmake version to 3.10 which is in line with the `CMakeLists` generated by `flutter create`. I think it's a bit annoying to require such a high version of cmake because it means developers have to jump through hoops to update it - e.g. on ubuntu 20.04, the highest you can get _easily_ is 3.16. Let me know if you're opposed to this change or if there are better ways to do it.

blagoev replied:
I think we should keep the required newer cmake version. We shouldn't be sticking to what `flutter create` does since these templates are not updated regularly but we indeed require newer cmake functionality. This is the same with cocoapods. 
If we need to make it even easier for developers we could update our readme with more details how to install these.

nirinchev replied:
Is that parameter that important though? Can we replace it with something and keep the functionality with the lower cmake version?

nirinchev replied:
As far as I can tell, this only matters if we're running multiple commands which we're not.

blagoev replied:
I have added a discussion point for this.

fealebenpae replied:
Can we have a flutter plugin without this native code that compiles at a user's machine? Our native code is pre-compiled as a shared library, so these dummy plugins do nothing of value.

nirinchev replied:
This downloads the native binary though, so we definitely need it.

fealebenpae replied:
I can't believe I'm saying this, but npm is better. In that case I agree with @nirinchev - it makes sense to stick to a lower CMake version, ideally the one that Flutter itself depends on, since we aren't actually compiling anything more complicated than the default plugin package template here.

nirinchev commented:
@blagoev I've addressed your comments and tested the S3 package as a path dependency on all supported platforms and it seems to work. This should be ready for a final review.

blagoev commented on flutter/realm_flutter/linux/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.19)
+cmake_minimum_required(VERSION 3.10)
```
we require Cmake 3.21 now. https://github.com/realm/realm-dart#environment-setup-for-realm-flutter
I think we should keep the requirement.

nirinchev replied:
This is different - we're not building Core here, so there's no reason to use 3.21.

fealebenpae replied:
Yeah, that's just the dummy flutter plugin - it doesn't build anything except the "hello world" plugin code. it doesn't need the newer CMake version.

blagoev replied:
yes indeed

blagoev commented:
@nirinchev the discussion could fix something in a folloupPR, so lets merge this one
